File: nlp/scribe_nlp/revision_ai.py
"""Dedicated local AI for document revision / diff analysis.

Heuristic, offline, no cloud LLM — classifies change kind, highlights risks,
and produces an editor-friendly report for the revision history panel.
"""

from __future__ import annotations

import re
from typing import Any, Literal

from .diff_summary import summarize_diff
from .text_utils import content_tokens, normalize_text, split_sentences

ChangeKind = Literal[
    "identical",
    "expansion",
    "trim",
    "rewrite",
    "polish",
    "structural",
    "mixed",
]

Severity = Literal["info", "warn", "critical"]

_HEADING_RE = re.compile(r"^(#{1,6}\s+|.+\n[=-]{3,}\s*$)", re.MULTILINE)
_MARKDOWN_HEADING_RE = re.compile(r"^(#{1,6})\s+(.+)$", re.MULTILINE)
_TODO_RE = re.compile(r"\b(TODO|FIXME|XXX|HACK|WIP)\b", re.IGNORECASE)
_SENSITIVE_RE = re.compile(
    r"\b(password|secret|api[_-]?key|token|ssn|rodn[eé]\s*[čc][ií]slo)\b",
    re.IGNORECASE,
)


def analyze_revision_diff(
    old_text: str,
    new_text: str,
    *,
    max_bullets: int = 6,
    language: str | None = None,
) -> dict[str, Any]:
    """Analyze a plain-text revision diff for the editor / Local AI panel."""
    old = old_text or ""
    new = new_text or ""
    max_bullets = max(1, min(int(max_bullets or 6), 12))

    base = summarize_diff(old, new, max_bullets=max_bullets)
    old_words = int(base.get("oldWordCount") or 0)
    new_words = int(base.get("newWordCount") or 0)
    change_ratio = float(base.get("changeRatio") or 0.0)
    added_sentences = list(base.get("addedSentences") or [])
    removed_sentences = list(base.get("removedSentences") or [])
    gained_terms = list(base.get("gainedTerms") or [])
    lost_terms = list(base.get("lostTerms") or [])

    old_lines = old.splitlines()
    new_lines = new.splitlines()
    lines_added, lines_removed = _line_churn(old_lines, new_lines)
    heading_changes = _heading_changes(old, new)
    risks = _detect_risks(old, new, old_words, new_words, change_ratio, lines_removed)

    change_kind = _classify_change_kind(
        old_words=old_words,
        new_words=new_words,
        change_ratio=change_ratio,
        lines_added=lines_added,
        lines_removed=lines_removed,
        heading_changes=heading_changes,
        added_n=len(added_sentences),
        removed_n=len(removed_sentences),
    )
    confidence = _confidence(change_kind, change_ratio, old_words, new_words)

    lang = (language or _guess_language(old + "\n" + new)).lower()
    headline = _headline(change_kind, old_words, new_words, change_ratio, lang)
    summary = _rich_summary(
        base_summary=str(base.get("summary") or ""),
        change_kind=change_kind,
        headline=headline,
        added=added_sentences,
        removed=removed_sentences,
        risks=risks,
        lang=lang,
    )
    bullets = _build_bullets(
        change_kind=change_kind,
        added=added_sentences,
        removed=removed_sentences,
        headings=heading_changes,
        risks=risks,
        gained=gained_terms,
        lost=lost_terms,
        max_bullets=max_bullets,
        lang=lang,
    )

    return {
        "summary": summary,
        "headline": headline,
        "changeKind": change_kind,
        "confidence": round(confidence, 3),
        "bullets": bullets,
        "addedSentences": added_sentences,
        "removedSentences": removed_sentences,
        "gainedTerms": gained_terms,
        "lostTerms": lost_terms,
        "headingChanges": heading_changes,
        "risks": risks,
        "stats": {
            "changeRatio": round(change_ratio, 3),
            "oldWordCount": old_words,
            "newWordCount": new_words,
            "linesAdded": lines_added,
            "linesRemoved": lines_removed,
            "netWords": new_words - old_words,
        },
        "source": "python",
        # Backward-compatible aliases for callers expecting summarize_diff shape.
        "changeRatio": round(change_ratio, 3),
        "oldWordCount": old_words,
        "newWordCount": new_words,
    }


def _line_churn(old_lines: list[str], new_lines: list[str]) -> tuple[int, int]:
    old_set = set(old_lines)
    new_set = set(new_lines)
    added = sum(1 for line in new_lines if line not in old_set)
    removed = sum(1 for line in old_lines if line not in new_set)
    return added, removed


def _extract_headings(text: str) -> list[str]:
    found: list[str] = []
    for match in _MARKDOWN_HEADING_RE.finditer(text or ""):
        title = normalize_text(match.group(2))
        if title:
            found.append(title)
    # TipTap plain export often has bare title lines without markdown hashes —
    # keep short ALL-CAPS / Title Case lines under 80 chars as weak headings.
    if not found:
        for line in (text or "").splitlines():
            stripped = line.strip()
            if 3 <= len(stripped) <= 80 and stripped == stripped.title() and " " in stripped:
                found.append(normalize_text(stripped))
    return found


def _heading_changes(old: str, new: str) -> dict[str, list[str]]:
    old_h = set(_extract_headings(old))
    new_h = set(_extract_headings(new))
    return {
        "added": sorted(new_h - old_h)[:8],
        "removed": sorted(old_h - new_h)[:8],
    }


def _detect_risks(
    old: str,
    new: str,
    old_words: int,
    new_words: int,
    change_ratio: float,
    lines_removed: int,
) -> list[str]:
    risks: list[str] = []
    if old_words > 40 and new_words < old_words * 0.55:
        risks.append("large_deletion")
    if change_ratio >= 0.65 and old_words > 80:
        risks.append("heavy_rewrite")
    if lines_removed >= 12:
        risks.append("many_lines_removed")
    old_todos = set(_TODO_RE.findall(old))
    new_todos = set(_TODO_RE.findall(new))
    if new_todos - old_todos:
        risks.append("todo_introduced")
    if _SENSITIVE_RE.search(new) and not _SENSITIVE_RE.search(old):
        risks.append("sensitive_term_added")
    if len(new.strip()) == 0 and len(old.strip()) > 0:
        risks.append("content_cleared")
    return risks


def _classify_change_kind(
    *,
    old_words: int,
    new_words: int,
    change_ratio: float,
    lines_added: int,
    lines_removed: int,
    heading_changes: dict[str, list[str]],
    added_n: int,
    removed_n: int,
) -> ChangeKind:
    if old_words == 0 and new_words == 0:
        return "identical"
    if change_ratio < 0.02 and lines_added == 0 and lines_removed == 0:
        return "identical"

    growth = (new_words - old_words) / max(old_words, 1)
    headings_moved = bool(heading_changes.get("added") or heading_changes.get("removed"))

    if headings_moved and abs(growth) < 0.25 and change_ratio < 0.45:
        return "structural"
    if growth >= 0.18 and change_ratio < 0.55:
        return "expansion"
    if growth <= -0.18 and change_ratio < 0.55:
        return "trim"
    if change_ratio >= 0.45 and abs(growth) < 0.2:
        return "rewrite"
    if change_ratio < 0.18 and abs(growth) < 0.12 and (added_n + removed_n) <= 4:
        return "polish"
    if abs(growth) >= 0.12 and change_ratio >= 0.35:
        return "mixed"
    if growth > 0.05:
        return "expansion"
    if growth < -0.05:
        return "trim"
    return "mixed"


def _confidence(
    change_kind: ChangeKind,
    change_ratio: float,
    old_words: int,
    new_words: int,
) -> float:
    if change_kind == "identical":
        return 0.99
    base = 0.55 + min(0.3, change_ratio)
    if old_words + new_words < 40:
        base -= 0.15
    return max(0.35, min(0.95, base))


def _guess_language(text: str) -> str:
    sample = (text or "")[:4000].lower()
    sk_hits = sum(
        1
        for token in (
            " že ",
            " nie ",
            " alebo ",
            " pre ",
            " ako ",
            " toto ",
            " ktoré ",
            " ktoré",
            "či ",
        )
        if token in sample
    )
    return "sk" if sk_hits >= 2 else "en"


def _headline(
    kind: ChangeKind,
    old_words: int,
    new_words: int,
    change_ratio: float,
    lang: str,
) -> str:
    pct = int(round(abs(new_words - old_words) / max(old_words, 1) * 100))
    if lang.startswith("sk"):
        labels = {
            "identical": "Bez zmien",
            "expansion": f"Rozšírenie (+{pct} % slov)",
            "trim": f"Skrátenie (−{pct} % slov)",
            "rewrite": f"Prepísanie (~{int(change_ratio * 100)} % zmeny)",
            "polish": "Jemná úprava",
            "structural": "Štrukturálna zmena (nadpisy)",
            "mixed": f"Zmiešaná úprava (~{int(change_ratio * 100)} %)",
        }
    else:
        labels = {
            "identical": "No changes",
            "expansion": f"Expansion (+{pct}% words)",
            "trim": f"Trim (−{pct}% words)",
            "rewrite": f"Rewrite (~{int(change_ratio * 100)}% changed)",
            "polish": "Light polish",
            "structural": "Structural (headings)",
            "mixed": f"Mixed edit (~{int(change_ratio * 100)}%)",
        }
    return labels.get(kind, labels["mixed"])


def _rich_summary(
    *,
    base_summary: str,
    change_kind: ChangeKind,
    headline: str,
    added: list[str],
    removed: list[str],
    risks: list[str],
    lang: str,
) -> str:
    parts = [headline]
    if base_summary and base_summary not in headline:
        parts.append(base_summary)
    elif added:
        parts.append(added[0])
    elif removed:
        prefix = "Odstránené" if lang.startswith("sk") else "Removed"
        parts.append(f"{prefix}: {removed[0]}")
    if "large_deletion" in risks:
        parts.append(
            "Pozor: veľká časť textu zmizla."
            if lang.startswith("sk")
            else "Warning: a large part of the text was removed."
        )
    if "todo_introduced" in risks:
        parts.append(
            "Pribudol TODO/FIXME marker."
            if lang.startswith("sk")
            else "A TODO/FIXME marker was introduced."
        )
    return " ".join(parts).strip()


def _build_bullets(
    *,
    change_kind: ChangeKind,
    added: list[str],
    removed: list[str],
    headings: dict[str, list[str]],
    risks: list[str],
    gained: list[str],
    lost: list[str],
    max_bullets: int,
    lang: str,
) -> list[dict[str, str]]:
    bullets: list[dict[str, str]] = []

    def push(text: str, severity: Severity, kind: str) -> None:
        if len(bullets) >= max_bullets or not text.strip():
            return
        bullets.append({"text": text.strip(), "severity": severity, "kind": kind})

    for risk in risks:
        push(_risk_label(risk, lang), "critical" if risk in {"content_cleared", "large_deletion"} else "warn", "risk")

    for title in headings.get("added", [])[:2]:
        label = "Nový nadpis" if lang.startswith("sk") else "New heading"
        push(f"{label}: {title}", "info", "heading")
    for title in headings.get("removed", [])[:2]:
        label = "Odstránený nadpis" if lang.startswith("sk") else "Removed heading"
        push(f"{label}: {title}", "warn", "heading")

    for sentence in added[:3]:
        push(sentence, "info", "added")
    for sentence in removed[:2]:
        push(sentence, "warn", "removed")

    if gained and len(bullets) < max_bullets:
        label = "Nové termíny" if lang.startswith("sk") else "New terms"
        push(f"{label}: {', '.join(gained[:5])}", "info", "terms")
    if lost and len(bullets) < max_bullets:
        label = "Stratené termíny" if lang.startswith("sk") else "Lost terms"
        push(f"{label}: {', '.join(lost[:5])}", "info", "terms")

    if not bullets and change_kind == "identical":
        push(
            "Texty sú rovnaké." if lang.startswith("sk") else "The texts are identical.",
            "info",
            "info",
        )
    return bullets


def _risk_label(risk: str, lang: str) -> str:
    sk = {
        "large_deletion": "Veľké vymazanie textu",
        "heavy_rewrite": "Silné prepísanie obsahu",
        "many_lines_removed": "Veľa odstránených riadkov",
        "todo_introduced": "Pribudol TODO/FIXME",
        "sensitive_term_added": "Možný citlivý výraz",
        "content_cleared": "Obsah bol úplne vymazaný",
    }
    en = {
        "large_deletion": "Large text deletion",
        "heavy_rewrite": "Heavy content rewrite",
        "many_lines_removed": "Many lines removed",
        "todo_introduced": "TODO/FIXME introduced",
        "sensitive_term_added": "Possible sensitive term added",
        "content_cleared": "Content was fully cleared",
    }
    table = sk if lang.startswith("sk") else en
    return table.get(risk, risk)

File: crates/scribe-core/src/nlp/revision_ai.rs
//! Dedicated local revision / diff AI (Rust fallback when Python NLP is off).

use crate::diff::diff_lines;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::sync::OnceLock;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum RevisionChangeKind {
    Identical,
    Expansion,
    Trim,
    Rewrite,
    Polish,
    Structural,
    Mixed,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RevisionAiBullet {
    pub text: String,
    pub severity: String,
    pub kind: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RevisionHeadingChanges {
    pub added: Vec<String>,
    pub removed: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RevisionAiStats {
    pub change_ratio: f64,
    pub old_word_count: i64,
    pub new_word_count: i64,
    pub lines_added: usize,
    pub lines_removed: usize,
    pub net_words: i64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RevisionAiReport {
    pub summary: String,
    pub headline: String,
    pub change_kind: RevisionChangeKind,
    pub confidence: f64,
    pub bullets: Vec<RevisionAiBullet>,
    pub added_sentences: Vec<String>,
    pub removed_sentences: Vec<String>,
    pub gained_terms: Vec<String>,
    pub lost_terms: Vec<String>,
    pub heading_changes: RevisionHeadingChanges,
    pub risks: Vec<String>,
    pub stats: RevisionAiStats,
    pub source: String,
    /// Compatibility with older summarize_diff consumers.
    pub change_ratio: f64,
    pub old_word_count: i64,
    pub new_word_count: i64,
}

fn heading_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(?m)^(#{1,6})\s+(.+)$").expect("heading regex"))
}

fn todo_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(?i)\b(TODO|FIXME|XXX|HACK|WIP)\b").expect("todo regex"))
}

fn word_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"[\w\u00C0-\u024F]{2,}").expect("word regex"))
}

fn sentence_split(text: &str) -> Vec<String> {
    let mut out = Vec::new();
    let mut buf = String::new();
    for ch in text.chars() {
        buf.push(ch);
        if matches!(ch, '.' | '!' | '?' | '…') {
            let trimmed = buf.trim().to_string();
            if !trimmed.is_empty() {
                out.push(trimmed);
            }
            buf.clear();
        }
    }
    let trimmed = buf.trim().to_string();
    if !trimmed.is_empty() {
        out.push(trimmed);
    }
    out
}

fn tokens(text: &str) -> HashSet<String> {
    word_re()
        .find_iter(text)
        .map(|m| m.as_str().to_lowercase())
        .filter(|w| w.len() > 2)
        .collect()
}

fn extract_headings(text: &str) -> HashSet<String> {
    heading_re()
        .captures_iter(text)
        .filter_map(|caps| caps.get(2).map(|m| m.as_str().trim().to_string()))
        .filter(|s| !s.is_empty())
        .collect()
}

fn guess_language(text: &str) -> &'static str {
    let sample = text.chars().take(4000).collect::<String>().to_lowercase();
    let sk_markers = [" že ", " nie ", " alebo ", " pre ", " ako ", " toto "];
    let hits = sk_markers.iter().filter(|m| sample.contains(*m)).count();
    if hits >= 2 {
        "sk"
    } else {
        "en"
    }
}

/// Analyze a revision diff offline (no Python sidecar required).
pub fn analyze_revision_diff(
    old_text: &str,
    new_text: &str,
    max_bullets: usize,
    language: Option<&str>,
) -> RevisionAiReport {
    let max_bullets = max_bullets.clamp(1, 12);
    let old = old_text;
    let new = new_text;
    let lang = language
        .map(|value| {
            if value.to_ascii_lowercase().starts_with("sk") {
                "sk"
            } else {
                "en"
            }
        })
        .unwrap_or_else(|| guess_language(&format!("{old}\n{new}")));

    let old_sentences = sentence_split(old);
    let new_sentences = sentence_split(new);
    let old_set: HashSet<_> = old_sentences.iter().cloned().collect();
    let new_set: HashSet<_> = new_sentences.iter().cloned().collect();
    let added_sentences: Vec<String> = new_sentences
        .iter()
        .filter(|s| !old_set.contains(*s))
        .cloned()
        .take(max_bullets)
        .collect();
    let removed_sentences: Vec<String> = old_sentences
        .iter()
        .filter(|s| !new_set.contains(*s))
        .cloned()
        .take(max_bullets)
        .collect();

    let old_tokens = tokens(old);
    let new_tokens = tokens(new);
    let mut gained: Vec<String> = new_tokens.difference(&old_tokens).cloned().collect();
    let mut lost: Vec<String> = old_tokens.difference(&new_tokens).cloned().collect();
    gained.sort();
    lost.sort();
    gained.truncate(12);
    lost.truncate(12);

    let union = old_tokens.union(&new_tokens).count();
    let change_ratio = if union == 0 {
        0.0
    } else {
        old_tokens.symmetric_difference(&new_tokens).count() as f64 / union as f64
    };

    let old_word_count = old.split_whitespace().count() as i64;
    let new_word_count = new.split_whitespace().count() as i64;

    let diff = diff_lines(old, new);
    let lines_added = diff.added;
    let lines_removed = diff.removed;

    let old_headings = extract_headings(old);
    let new_headings = extract_headings(new);
    let mut heading_added: Vec<String> = new_headings.difference(&old_headings).cloned().collect();
    let mut heading_removed: Vec<String> = old_headings.difference(&new_headings).cloned().collect();
    heading_added.sort();
    heading_removed.sort();
    heading_added.truncate(8);
    heading_removed.truncate(8);

    let mut risks = Vec::new();
    if old_word_count > 40 && new_word_count < ((old_word_count as f64) * 0.55) as i64 {
        risks.push("large_deletion".into());
    }
    if change_ratio >= 0.65 && old_word_count > 80 {
        risks.push("heavy_rewrite".into());
    }
    if lines_removed >= 12 {
        risks.push("many_lines_removed".into());
    }
    if todo_re().is_match(new) && !todo_re().is_match(old) {
        risks.push("todo_introduced".into());
    }
    if new.trim().is_empty() && !old.trim().is_empty() {
        risks.push("content_cleared".into());
    }

    let change_kind = classify(
        old_word_count,
        new_word_count,
        change_ratio,
        !heading_added.is_empty() || !heading_removed.is_empty(),
        added_sentences.len(),
        removed_sentences.len(),
        lines_added,
        lines_removed,
    );
    let confidence = confidence(change_kind, change_ratio, old_word_count, new_word_count);
    let headline = headline(change_kind, old_word_count, new_word_count, change_ratio, lang);
    let summary = rich_summary(
        &headline,
        &added_sentences,
        &removed_sentences,
        &risks,
        lang,
    );
    let bullets = build_bullets(
        change_kind,
        &added_sentences,
        &removed_sentences,
        &heading_added,
        &heading_removed,
        &risks,
        &gained,
        &lost,
        max_bullets,
        lang,
    );

    RevisionAiReport {
        summary,
        headline,
        change_kind,
        confidence: (confidence * 1000.0).round() / 1000.0,
        bullets,
        added_sentences,
        removed_sentences,
        gained_terms: gained,
        lost_terms: lost,
        heading_changes: RevisionHeadingChanges {
            added: heading_added,
            removed: heading_removed,
        },
        risks,
        stats: RevisionAiStats {
            change_ratio: (change_ratio * 1000.0).round() / 1000.0,
            old_word_count,
            new_word_count,
            lines_added,
            lines_removed,
            net_words: new_word_count - old_word_count,
        },
        source: "rust".into(),
        change_ratio: (change_ratio * 1000.0).round() / 1000.0,
        old_word_count,
        new_word_count,
    }
}

fn classify(
    old_words: i64,
    new_words: i64,
    change_ratio: f64,
    headings_moved: bool,
    added_n: usize,
    removed_n: usize,
    lines_added: usize,
    lines_removed: usize,
) -> RevisionChangeKind {
    if old_words == 0 && new_words == 0 {
        return RevisionChangeKind::Identical;
    }
    if change_ratio < 0.02 && lines_added == 0 && lines_removed == 0 {
        return RevisionChangeKind::Identical;
    }
    let growth = (new_words - old_words) as f64 / old_words.max(1) as f64;
    if headings_moved && growth.abs() < 0.25 && change_ratio < 0.45 {
        return RevisionChangeKind::Structural;
    }
    if growth >= 0.18 && change_ratio < 0.55 {
        return RevisionChangeKind::Expansion;
    }
    if growth <= -0.18 && change_ratio < 0.55 {
        return RevisionChangeKind::Trim;
    }
    if change_ratio >= 0.45 && growth.abs() < 0.2 {
        return RevisionChangeKind::Rewrite;
    }
    if change_ratio < 0.18 && growth.abs() < 0.12 && (added_n + removed_n) <= 4 {
        return RevisionChangeKind::Polish;
    }
    if growth.abs() >= 0.12 && change_ratio >= 0.35 {
        return RevisionChangeKind::Mixed;
    }
    if growth > 0.05 {
        RevisionChangeKind::Expansion
    } else if growth < -0.05 {
        RevisionChangeKind::Trim
    } else {
        RevisionChangeKind::Mixed
    }
}

fn confidence(kind: RevisionChangeKind, change_ratio: f64, old_words: i64, new_words: i64) -> f64 {
    if kind == RevisionChangeKind::Identical {
        return 0.99;
    }
    let mut base = 0.55 + change_ratio.min(0.3);
    if old_words + new_words < 40 {
        base -= 0.15;
    }
    base.clamp(0.35, 0.95)
}

fn headline(
    kind: RevisionChangeKind,
    old_words: i64,
    new_words: i64,
    change_ratio: f64,
    lang: &str,
) -> String {
    let pct = (((new_words - old_words).abs() as f64 / old_words.max(1) as f64) * 100.0).round() as i64;
    let ratio_pct = (change_ratio * 100.0).round() as i64;
    let sk = lang.starts_with("sk");
    match kind {
        RevisionChangeKind::Identical => {
            if sk {
                "Bez zmien".into()
            } else {
                "No changes".into()
            }
        }
        RevisionChangeKind::Expansion => {
            if sk {
                format!("Rozšírenie (+{pct} % slov)")
            } else {
                format!("Expansion (+{pct}% words)")
            }
        }
        RevisionChangeKind::Trim => {
            if sk {
                format!("Skrátenie (−{pct} % slov)")
            } else {
                format!("Trim (−{pct}% words)")
            }
        }
        RevisionChangeKind::Rewrite => {
            if sk {
                format!("Prepísanie (~{ratio_pct} % zmeny)")
            } else {
                format!("Rewrite (~{ratio_pct}% changed)")
            }
        }
        RevisionChangeKind::Polish => {
            if sk {
                "Jemná úprava".into()
            } else {
                "Light polish".into()
            }
        }
        RevisionChangeKind::Structural => {
            if sk {
                "Štrukturálna zmena (nadpisy)".into()
            } else {
                "Structural (headings)".into()
            }
        }
        RevisionChangeKind::Mixed => {
            if sk {
                format!("Zmiešaná úprava (~{ratio_pct} %)")
            } else {
                format!("Mixed edit (~{ratio_pct}%)")
            }
        }
    }
}

fn rich_summary(
    headline: &str,
    added: &[String],
    removed: &[String],
    risks: &[String],
    lang: &str,
) -> String {
    let mut parts = vec![headline.to_string()];
    if let Some(first) = added.first() {
        parts.push(first.clone());
    } else if let Some(first) = removed.first() {
        let prefix = if lang.starts_with("sk") {
            "Odstránené"
        } else {
            "Removed"
        };
        parts.push(format!("{prefix}: {first}"));
    }
    if risks.iter().any(|r| r == "large_deletion") {
        parts.push(if lang.starts_with("sk") {
            "Pozor: veľká časť textu zmizla.".into()
        } else {
            "Warning: a large part of the text was removed.".into()
        });
    }
    parts.join(" ")
}

fn build_bullets(
    _kind: RevisionChangeKind,
    added: &[String],
    removed: &[String],
    heading_added: &[String],
    heading_removed: &[String],
    risks: &[String],
    gained: &[String],
    lost: &[String],
    max_bullets: usize,
    lang: &str,
) -> Vec<RevisionAiBullet> {
    let mut bullets = Vec::new();
    let sk = lang.starts_with("sk");

    let push = |bullets: &mut Vec<RevisionAiBullet>, text: String, severity: &str, kind: &str| {
        if bullets.len() >= max_bullets || text.trim().is_empty() {
            return;
        }
        bullets.push(RevisionAiBullet {
            text,
            severity: severity.into(),
            kind: kind.into(),
        });
    };

    for risk in risks {
        let severity = if matches!(risk.as_str(), "content_cleared" | "large_deletion") {
            "critical"
        } else {
            "warn"
        };
        push(&mut bullets, risk_label(risk, sk), severity, "risk");
    }
    for title in heading_added.iter().take(2) {
        let label = if sk { "Nový nadpis" } else { "New heading" };
        push(&mut bullets, format!("{label}: {title}"), "info", "heading");
    }
    for title in heading_removed.iter().take(2) {
        let label = if sk {
            "Odstránený nadpis"
        } else {
            "Removed heading"
        };
        push(&mut bullets, format!("{label}: {title}"), "warn", "heading");
    }
    for sentence in added.iter().take(3) {
        push(&mut bullets, sentence.clone(), "info", "added");
    }
    for sentence in removed.iter().take(2) {
        push(&mut bullets, sentence.clone(), "warn", "removed");
    }
    if !gained.is_empty() && bullets.len() < max_bullets {
        let label = if sk { "Nové termíny" } else { "New terms" };
        push(
            &mut bullets,
            format!("{label}: {}", gained.iter().take(5).cloned().collect::<Vec<_>>().join(", ")),
            "info",
            "terms",
        );
    }
    if !lost.is_empty() && bullets.len() < max_bullets {
        let label = if sk { "Stratené termíny" } else { "Lost terms" };
        push(
            &mut bullets,
            format!("{label}: {}", lost.iter().take(5).cloned().collect::<Vec<_>>().join(", ")),
            "info",
            "terms",
        );
    }
    bullets
}

fn risk_label(risk: &str, sk: bool) -> String {
    match (risk, sk) {
        ("large_deletion", true) => "Veľké vymazanie textu".into(),
        ("large_deletion", false) => "Large text deletion".into(),
        ("heavy_rewrite", true) => "Silné prepísanie obsahu".into(),
        ("heavy_rewrite", false) => "Heavy content rewrite".into(),
        ("many_lines_removed", true) => "Veľa odstránených riadkov".into(),
        ("many_lines_removed", false) => "Many lines removed".into(),
        ("todo_introduced", true) => "Pribudol TODO/FIXME".into(),
        ("todo_introduced", false) => "TODO/FIXME introduced".into(),
        ("content_cleared", true) => "Obsah bol úplne vymazaný".into(),
        ("content_cleared", false) => "Content was fully cleared".into(),
        (other, _) => other.into(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn expansion_report() {
        let old = "Meeting notes. We agreed on the timeline.";
        let new = "Meeting notes. We agreed on the timeline. Next steps include drafting the proposal and scheduling a review for Friday.";
        let report = analyze_revision_diff(old, new, 6, Some("en"));
        assert_eq!(report.source, "rust");
        assert!(matches!(
            report.change_kind,
            RevisionChangeKind::Expansion | RevisionChangeKind::Mixed | RevisionChangeKind::Polish
        ));
        assert!(!report.headline.is_empty());
    }

    #[test]
    fn identical_report() {
        let report = analyze_revision_diff("Same", "Same", 4, Some("en"));
        assert_eq!(report.change_kind, RevisionChangeKind::Identical);
    }
}
